### PR TITLE
Speed up DWG writer

### PR DESCRIPTION
**Pull request by gongxy-sh**
Before：
100000 entities used 00:03:34
300000 entities used 00:33:53

After:
100000 entities used 00:00:02
300000 entities used sed 00:00:23

### CODE REVIEW

**Comment by gongxy-sh** [comment]
Main code as 
Dwg_Data* pDoc = dwg_new_Document(R_2000, 0, 0);
  Dwg_Object* poMSpace = dwg_model_space_object(pDoc);
  Dwg_Object_BLOCK_HEADER* pMSpace = poMSpace->tio.object->tio.BLOCK_HEADER;
  dwg_point_3d start = { 0 };
  dwg_point_3d end = { 0 };
  for(size_t i = 0; i < ....;i++)
  {
    end.x = 10 * i;
    end.y = 10 * i;
    dwg_add_LINE(pMSpace, &start, &end);
  }
  dwg_write_file(..., pDoc);
  dwg_free_Document(pDoc);

File: src/dwg.c [hunk 1]
```diff
@@ -3192,3 +3205,230 @@ dwg_log_dataflags (const int _loglevel, const int maxlevel,
         HANDLER (OUTPUT, "  vert_alignment (128)\n");
     }
 }
+
+static int
+Dwg_Object_Ref_compare (const Dwg_Object_Ref *key, const Dwg_Object_Ref **ppR)
```
Comment (rurban) [inline src/dwg.c:3210]: static methods don't need the `Dwg_` prefix. `ref_cmp` is enough.

File: src/dwg.c [hunk 2]
```diff
@@ -3192,3 +3205,230 @@ dwg_log_dataflags (const int _loglevel, const int maxlevel,
...
+ *           returns NULL
+ *           *ppBefore pointer to key's insert pos in array
+ *
+ *Exceptions:
+ *   Input parameters are validated. Refer to the validation section of the
+ *function.
+ *
+ *******************************************************************************/
+static void *
+bsearch_ex (const void *key, const void *base, size_t num, size_t width,
```
Comment (rurban) [inline src/dwg.c:3253]: bsearch is in C99 already. Is this needed for VS2019?

File: src/dwg.c (hunk 1 above)
Comment (gongxy-sh) [reply src/dwg.c:3210]: OK. Thanks.

File: src/dwg.c [hunk 3]
```diff
@@ -3192,3 +3205,230 @@ dwg_log_dataflags (const int _loglevel, const int maxlevel,
...
+        {
+          // key < lo
+          *ppBefore = lo;
+        }
+    }
+  return NULL;
+}
+
+void
+dwg_ordered_ref_appended (Dwg_Data *restrict dwg, Dwg_Object_Ref *ref)
```
Comment (rurban) [inline src/dwg.c:3328]: maybe just dwg_ordered_object_refs_add(). It's either insert or append

File: src/dwg.c (hunk 2 above)
Comment (gongxy-sh) [reply src/dwg.c:3253]: I need bsearch_ex to return the insertion point when the key is not found.

File: src/dwg.c (hunk 3 above)
Comment (gongxy-sh) [reply src/dwg.c:3328]: OK.

File: src/encode.c [hunk 4]
```diff
@@ -5766,6 +5766,7 @@ find_prev_entity (Dwg_Object *obj)
   Dwg_Data *dwg = obj->parent;
   if (obj->supertype != DWG_SUPERTYPE_ENTITY)
     return NULL;
+  /*
   for (BITCODE_BL i = obj->index - 1; i > 0; i--)
```
Comment (rurban) [inline src/encode.c:5770]: They are already searched backwards. Are there really so many non-entities? well, there are at the end, but still curious about the speedup.

File: src/dwg.c (hunk 2 above)
Comment (rurban) [reply src/dwg.c:3253]: Oh my, well then

File: src/dwg_api.c [hunk 5]
```diff
@@ -23063,6 +23083,11 @@ dwg_insert_entity (Dwg_Object_BLOCK_HEADER *restrict _owner,
         }
     }
   in_postprocess_handles (obj);
+  //for speed up find_prev_entity
+  if (DWG_SUPERTYPE_ENTITY == obj->fixedtype)
+    {
+      dwg->prev_entity = obj->index;
```
Comment (rurban) [inline src/dwg_api.c:23089]: => prev_entity_index please.

Comment (gongxy-sh) [reply src/dwg_api.c:23089]: OK

File: src/encode.c (hunk 4 above)
Comment (gongxy-sh) [reply src/encode.c:5770]: lsearch
100000 entities used 00:03:34
300000 entities used 00:33:53

bsearch
100000 entities Used 00:01:19
300000 entities used  00:14:53

bsearch+prev_entity

100000 entities used 00:00:02
300000 entities used 00:00:23

**Comment by rurban** [comment]
No, no extra memory management isolation functions. 
For free() we do have FREE_IF, the rest is standardized and could be overridden by boem-gc/jemalloc/tcmalloc/mimalloc/... headers.

**Comment by gongxy-sh** [comment]
The following scenario: dwg writer, app debug mode/dwg library release mode

Dwg_ Data* dwg = dwg_ new_ Document(...);
...
dwg_ free(dwg);
free(dwg); -- This will cause crash, and memory leakage if not released

**Comment by gongxy-sh** [comment]
By memory management isolation, I mean that whoever applies for memory will release it. In other words, the malloc/calloc/realloc/free functions called by the application and the dwg library may not be the same set of functions.

**Pull request closed without merging**